RFP-01.md

```diff
@@ -0,0 +1,169 @@
+     * @dev Emits an event when setting a delay
+     * @param delay The delay in seconds
+     */
+    function setExitDelay(uint256 delay) external auth;
+    /**
+     * Allow `token` to be used as collateral for the sub-pool.
+     * @dev Emits an event `AllowCollateral` when allowing/updating a token
+     */
+    function allowCollateral(IERC20 token, uint256 min) external auth;
+    /**
+     * Revoke `token` as a collateral for sub-pools
+     * @dev Emits an event `RevokeCollateral` when revoking a token
+     */
+    function revokeCollateral(IERC20 token) external auth;
+    /**
+     * Fine a sub-pool
+     * @param pool Which sub-pool to fine
+     * @param tokens The tokens to fine (sending to `msg.sender`)
+     * @param amts The amount of tokens to fine
+     * @dev Emits an event `Fine` when fining a sub-pool
+     */
+    function fine(SubPool pool, IERC20[] calldata tokens, uint256[] calldata amts) external auth;
+
+    // --- SubPool authed functions ---
+
+    /**
+     * Indicate a sub-pool's intent to quit the batch auction
+     * At this point the solver whose sub-pool it is is now removed from
+     * the competition.
+     * This then starts the `delay` timer until exit is allowed.
+     * @param pool The sub-pool to announce an exit for
+     * @dev Emits an event `AnnounceExit`
+     */
+    function announceDestroy(SubPool pool) external onlySubpoolAuth;
+    /**
+     * Withdraw any remaining collateral / cow from the sub-pool.
+     * Funds are sent to `msg.sender`.
+     * @dev Should emit an event `SubPoolDestroyed` on the factory
+     */
+    function finalizeExit(SubPool pool) external onlySubpoolAuth;
+    /**
+     * Set the backend URI for the sub-pool
+     * @param pool The sub-pool to set the URI for
+     * @param newUri The new URI to set
+     * @dev Emits an event `SetBackendUri` on the factory when setting a URI
+     */
+    function setBackendUri(SubPool pool, string calldata newUri) external onlySubpoolAuth;
+}
+
+interface SignedSettlement is Auth {
+    /**
+     * Signed settlement function that is callable by `msg.sender` if `msg.sender` is in
+     * receipt of a valid signature. The signer is a trusted entity and bears the responsibility
+     * of ensuring that `msg.sender` is correctly collateralised.
+     *  
+     * The signed data **MUST** include:
+     * - `solver` - the address of the sub-bonded solver, i.e. `msg.sender`
+     * - Replay protection - an expiry based on some configurable method (e.g. block number)
+     *
+     * @notice This contract is a facade around the `GPv2Settlement` contract and has no
+     *         on-chain state / binding to the concept of a sub-pool. It is merely a
+     *         permissioned wrapper around the `settle` function.
+     * @dev There must be **NO** use of `SLOAD` / `SSTORE` in the execution path.
+     */
+    function signedSettle(
+        bytes calldata signedData,
+        bytes calldata unsignedData,
+    ) external onlySubPool;
```

Comment:
@fleupold

Reply:
This decision is key and should be included in the document. This is what I think should be included in the signature:

```solidity
struct SignatureParts {
  /// The solver address that should be checked against `msg.sender``.
  address solver;
  /// The signature is invalid after this time.
  /// Note that signatures can be replayed until the deadline passes.
  uint32 deadline;
  // Parameters from the call to `settle`
  IERC20[] calldata tokens;
  uint256[] calldata clearingPrices;
  GPv2Trade.Data[] calldata trades;
}
```

This should not be, though it will be included in the function input in some form:
```solidity
struct NotSigned {
  // Parameters from the call to `settle`
  GPv2Interaction.Data[][3] calldata interactions;
}
```

I wrote them as Solidity struct for clarity but they don't have to be, I expect them to be packed in some reasonable format.

Reply:
I don't think you will be able to hash in-place for 712 because each element of the struct will need to be hashed and then hash the hashes, etc? But yes, we should reuse the abi encoding instead of encoding the args again since its the same structures and order of args, just need to patch a selector at the start.

Reply:
Hmm, i just noticed you mentioned eth-sign so assuming 191 we can do minimal memory manipulation to get the digest to sign indeed 👍.

Reply:
I don't think we should be going into _concrete_ implementation details here (i.e. yul suggestions etc). This is for a grant afterall (and they're paid for this :laughing:)

Having said this - what needs to be signed is important. I agree with the `SignatureParts` struct suggestion, however @fleupold made mention of the interactions - notably there should be some binding restriction that a user's pre and post interactions are included (so CoW DAO bonding pool meets it's social consensus obligations). So this muddies it a bit as to do:

1. Just pre/post; or
2. All interactions

The outcome of the above I think would likely rest in gas profiling, so I think being overly specific as to how the implementation should work isn't what we should do here and just kick it back to the grantee.

Reply:
Yeah, the only part that should appear in the document is details on what has to be signed, I got carried away. :slightly_smiling_face: 
A takeaway from this was that for something that needs to be gas efficient, we should probably not specify an interface at all but just say what we need. Then the actual implementation can be optimized without having to fit an interface. It's a different thing for the other contracts where gas efficiency is less of a concern.

Reply:
if we using eth-sign aka 191 including interactions in the signature is trivial, for 721 you need to do more stuff and more hashes gas wise. Of course with number of interactions and size of interactions the gas for hash will scale linearly too.

Reply:
The main reason why we wanted to avoid signing interactions in the first place is gas efficiency, less things to hash. On the other hand, the original concern was that we'd have to write to memory twice, whose cost grows quadratically with size. Writing once is unavoidable because of the call, but if we don't need to write to memory more then we only need to work with the cost of extra hashing, 6 bytes per word.

In the last 20 transactions, I see that this would be more or less between 207 and 669 extra gas, which I think is acceptable. So maybe we should hash everything.

<details><summary>analysis of random transactions</summary>

["Lean" transaction](https://etherscan.io/tx/0x649295f62bac4c971c5564830c44cff32d66dde77772adc383317bd3e2af706c) (just Uniswapv2)
Calldata bytes: 2092
Bytes dedicated to everything but interactions: 0x3e0 = 992
Cost of extra hashing: 6 · ceil((209 - 992)/32) = **207**

["Fattest" transaction](https://etherscan.io/tx/0xd7f21343c59f1b215e84d0c0b29462ad29779b4c364f7808c04ceb0488991a90) among the most recent 20.
Calldata bytes: 4556
Bytes dedicated to everything but interactions: 0x3e0 = 992, again
Cost of extra hashing: 6 · ceil((4556 - 992)/32) = **669**

</details>

Reply:
>   /// The signature is invalid after this time.
  /// Note that signatures can be replayed until the deadline passes.
  uint32 deadline;
  
 Is there a specific reason to make this a timestamp instead of a block? If it were a block there would be less uncertainty whether it still makes sense for the solver to submit the solution. The less predictable the block times of a chain are the less clear the submission strategy becomes IMO.
 Additionally this would already fit nicely into a possible 1 block per auction world.
 
Why should interactions not be part of the commitment? I think one of the warts of the current contract is that there are no guarantees around pre- and post-interactions. Verifying these would resolve this, no?

Reply:
If possible, we shouldn't require signing all interactions. The `driver` doesn't report their whole solution back to the autopilot, therefore the autopilot cannot sign off on all pre/post interactions a solver may want to include (e.g. for just-in-time CoW AMM orders). While we could change this, this would require a larger update to the existing driver<>autopilot communication.

What the autopilot does know is which user orders are going to be settled and whether those have `pre` or `post` interactions (available in the orderbook DB).

Ideally it would commit to just this subset of interactions, so that the wrapper contract can enforce inclusion of user specified pre and post interactions.

Reply:
> If possible, we shouldn't require signing all interactions. The driver doesn't report their whole solution back to the autopilot, therefore the autopilot cannot sign off on all pre/post interactions a solver may want to include (e.g. for just-in-time CoW AMM orders). While we could change this, this would require a larger update to the existing driver<>autopilot communication.

For this, I think we should be governed by gas considerations here. Let's make it part of the spec that the grantee should assess both methods for gas optimisation, and then make a decision on the analysis.